mmv1/third_party/terraform/services/container/node_config.go.erb

```diff
@@ -78,6 +78,15 @@ func schemaContainerdConfig() *schema.Schema {
 	}
 }
 
+func schemaInsecureKubeletReadonlyPortEnabled() *schema.Schema {
+	return &schema.Schema{
+		Type: schema.TypeBool,
+		Optional: true,
+		Computed: true,
```

Comment:
@trodge: do you have thoughts on whether this should be true (here and in the other spot(s) this is used)?

Reply:
Got it. Will give that a shot. Thanks for the additional guidance.

Just to confirm, this mean the value in the config settings will be `ENABLED` or `DISABLED` as well, correct vs. just using this internally in the state? I think get the idea, though IMO, having it be a bool is a lot easier and more consistent from an end-user perspective, and possibly more consistent with other params in this resource. It's also going to be quite a big update to the PR, so I want to make sure I get that part right. It's also a bit ugly if we leave `_enabled` on the parameter name; should we go from
```hcl
insecure_kubelet_readonly_port_enabled = true
```
to
```hcl
insecure_kubelet_readonly_port = "ENABLED" ## or "DISABLED"
```

I had just been working on adding `ForceSendFields` in various spots, though hadn't gotten it fixed yet.

Reply:
ps - If you're able to rework it more easily than all the back and forth getting it updated to the above will entail, totally fine with that as well, though happy to make a pass at it if you want me to update it.

Reply:
Leaving as a boolean will be problematic if the API default ever changes, because users will be unable to distinguish between leaving it unspecified and explicitly setting it to false. This is the approach we've taken to similar situations in the past, such as in [org_policy_policy](https://github.com/hashicorp/terraform-provider-google/blob/479c901bf023bfcca8648358e24d3a795828afb3/google/services/orgpolicy/resource_org_policy_policy.go#L137-L141). In that case, all three of `true`, `false`, and `nil` were valid values from the API's perspective. This situation is a little bit different in that the latest version of the API only supports `true` and `false`, but similar in the sense that we don't want to invalidate configs that don't specify a value. Using an enum is the only way to distinguish between users setting no value (as they would in versions where the resource doesn't exist) and setting it to `false`. The API can then decide which default is appropriate when the user indicates no preference.

I agree that

```
insecure_kubelet_readonly_port_enabled = "DISABLED"
```
is undesirable. Either of 
```
insecure_kubelet_readonly_port = "ENABLED" ## or "DISABLED"
```
or
```
insecure_kubelet_readonly_port_enabled = "TRUE" ## or "FALSE"
```

seem clear to me, but I'm not really a frequent user of the resource so I'm not sure which one is the most readable.

I would love for you to keep working on this, since you seem to have a better handle on the complexity of these hand-written resources. However, if you'd rather I pick it up from here I can give it a try.

Reply:
```
insecure_kubelet_readonly_port_enabled = "TRUE" ## or "FALSE"
```

In that case, the enum would be `TRUE` / `FALSE` vs `ENABLED` / `DISABLED`, right?
I tend to prefer that, because it would hew a little closer to the underlyling APIs, even if it's not a true Terraform boolean (I think it would also simplify the adjustments to the code at this point, though either way there's a whole lot of stuff that needs to get updated). There seems to be some prior art in both directions in the codebase. Can you doublecheck with whoever on the team had an opinion on this to make sure they're good with using `TRUE` / `FALSE` here?


(looks like for the hand written resources, the enum looks like `ValidateFunc: validation.StringInSlice([]string{"DISABLED","ENABLED"}, false),`, correct?)

Reply:
Double checked with my tech lead and we're in agreement that enum is the way to go and `"TRUE"`/`"FALSE"` make sense as values to align with the API.

And yes, using `validation.StringInSlice` as seen elsewhere will be ideal here.

Reply:
Great.
> Using an enum is the only way to distinguish between users setting no value (as they would in versions where the resource doesn't exist) and setting it to false.

So are we allowing `""` in the validation as well to represent the default, if I'm understanding you correctly, right?  Or just TRUE / FALSE only?

The flattener would then look like this?

```go
func flattenInsecureKubeletReadonlyPortEnabled(c *container.NodeKubeletConfig) string {
  insecureKubeletReadonlyPortEnabled := ""
  if c != nil && c.InsecureKubeletReadonlyPortEnabled {
    insecureKubeletReadonlyPortEnabled = "TRUE"
  }
  return insecureKubeletReadonlyPortEnabled
}
```

Reply:
> The flattener should return "ENABLED" when the API value is true and "DISABLED" otherwise.

Just want to make sure I'm not getting tangled up here about what's going in which direction... the flattener for, e.g., `KubeletConfig` is subclassed to `container.NodeKubeletConfig`, so if I try to do something like this:

```go
func flattenKubeletConfig(c *container.NodeKubeletConfig) []map[string]interface{} {
[...]
      "insecure_kubelet_readonly_port_enabled": flattenInsecureKubeletReadonlyPortEnabled(c.InsecureKubeletReadonlyPortEnabled),
```

(where `flattenInsecureKubeletReadonlyPortEnabled` returns a string), I'll get an error like:
```
cannot use c.InsecureKubeletReadonlyPortEnabled (variable of type bool) as *"google.golang.org/api/container/v1".NodeKubeletConfig value in argument to flattenInsecureKubeletReadonlyPortEnabled
```

I can push up my non-working updates if that's helpful.

Reply:
Pushed up the current progress, even though there are obvious failures... can you let me know if things are going in the right direction?

Reply:
@trodge I have things now in a partially working(ish) state, still some issues with `node_config.kubelet_config` on updates, I think. I note with https://github.com/hashicorp/terraform-provider-google/issues/15767 that there are some existing quirks with some of those parameters, but not sure if there's any connection or if I've just got to play with the forcesendfield settings a bit more.

```
--- PASS: TestAccContainerCluster_withInsecureKubeletReadonlyPortEnabledInNodeConfig (554.66s)
--- PASS: TestAccContainerCluster_withInsecureKubeletReadonlyPortEnabledInNodePool (573.91s)
--- PASS: TestAccContainerCluster_withInsecureKubeletReadonlyPortEnabledUpdates (606.97s)
--- PASS: TestAccContainerNodePool_withKubeletConfig (1063.76s)
```

I've done my best to keep stuff updated, but would be great if, once we get all the other stuff squared away, you can also double-check that the go and ruby templates are still identical (including whitespace) and that the docs updates are correct etc.